romancal/flux/flux_step.py

```diff
@@ -0,0 +1,114 @@
+        either an association file or wrapped by a
+        `~romancal.datamodels.container.ModelContainer`.
+
+    Returns
+    -------
+    output_models : `roman_datamodels.datamodels.DataModel`, or `~romancal.datamodels.container.ModelContainer`
+        The models with flux applied.
+
+
+    Notes
+    -----
+    Currently, the correction is done in-place; the inputs are directly modified if in-memory DataModels are input.
+    """  # noqa: E501
+
+    class_alias = "flux"
+
+    spec = """
+    """  # noqa: E501
+
+    reference_file_types = []
+
+    def process(self, input):
+        if isinstance(input, datamodels.DataModel):
+            input_models = ModelContainer([input])
+            single_model = True
+        elif isinstance(input, str):
+            # either a single asdf filename or an association filename
+            try:
+                # association filename
+                input_models = ModelContainer(input)
+                single_model = False
+            except Exception:
+                # single ASDF filename
+                input_models = ModelContainer([input])
+                single_model = True
+        elif isinstance(input, ModelContainer):
+            input_models = input
+            single_model = False
+        else:
+            raise TypeError(
+                "Input must be an ASN filename, a ModelContainer, "
+                "a single ASDF filename, or a single Roman DataModel."
+            )
+
+        for model in input_models:
+            apply_flux_correction(model)
+
+        if single_model:
+            return input_models[0]
+        return input_models
+
+
+def apply_flux_correction(model):
+    """Apply the flux correction
+
+    The input model is expected to be Roman ImageModel-like.
+    The photometry information is taken from `meta.photometry`.
+
+    The model is converted in-place.
+
+    Parameters
+    ----------
+    model : `ImageModel`
+        The model to apply the flux correction to. The model is modified in-place.
+
+    Notes
+    -----
+    The modifications to the model can result in validation issues due to change of units.
+    """
+    # Define the various variance arrays
+    VARIANCES = ("var_rnoise", "var_poisson", "var_flat")
+
+    # Check for units. Must be election/second. Otherwise, it is unknown how to
+    # convert.
+    if model.data.unit != u.electron / u.s:
+        log.debug(
+            "Input data is not in units of e/s. Flux correction will not be done."
+        )
+        log.debug("Input data units are %s", model.data.unit)
+        return
+
+    # Apply the correction
+    # Assignments into the model are done through `_instance` to avoid
+    # validation errors on the units.
+    log.debug("Flux correction being applied")
+    c_mj = model.meta.photometry.conversion_megajanskys
+    model._instance["data"] = model.data * c_mj
+    for variance in VARIANCES:
+        model._instance[variance] = getattr(model, variance) * c_mj**2
```

Comment:
The 'err' column also needs to get multiplied by c_mj.

The model._instance[field] is to avoid validation, which we expect will not be needed if we loosen the set of allowed units in the wfi-image schema?  It will be challenging for us to save these at present?

Reply:
Oh, and I haven't paid attention to whether conversion_megajanskys carries units or not; if it's just a scalar, we want to update the units.

Reply:
The `_instance` use here was really because of a bug in roman_datamodels that has been resolved by PR#322. After pulling roman_datamodels:main, the use of `_instance` is removed.

`conversion_magajanskys` does carry units. Oddly errors are not raised at this point, though I suspect there will be when models are serialized.  Regardless, the aforementioned widening of allowed units will resolve.